src/main/java/com/hierynomus/smbj/share/Directory.java

```diff
@@ -126,56 +73,87 @@ public String toString() {
-                if (ctx.entries == null || ctx.entries.size() == 0 || ctx.lastStatus == NtStatus.STATUS_NO_MORE_FILES) {
-                    if (ctx.iteratorIndex > 0) {
-                        ctx.next = null;
-                    }
-                } else if (ctx.iteratorIndex == ctx.entries.size() - 1) {
-                    ctx.next = ctx.entries.get(ctx.entries.size() - 1);
-                    ctx.iteratorIndex = 0;
-                    query(ctx);
+        private FileInfo prepareNext() {
+            while (currentIterator != null) {
+                if (currentIterator.hasNext()) {
+                    return currentIterator.next();
                 } else {
-                    ctx.next = ctx.entries.get(ctx.iteratorIndex);
+                    currentIterator = queryDirectory(false);
                 }
             }
+            return null;
+        }
+
+        private Iterator<FileInfo> queryDirectory(boolean firstQuery) {
+            Session session = treeConnect.getSession();
+            Connection connection = session.getConnection();
+
+            // Query Directory Request
+            EnumSet<SMB2QueryDirectoryRequest.SMB2QueryDirectoryFlags> flags;
+            if (firstQuery) {
+                flags = EnumSet.of(SMB2QueryDirectoryRequest.SMB2QueryDirectoryFlags.SMB2_RESTART_SCANS);
+            } else {
+                flags = EnumSet.noneOf(SMB2QueryDirectoryRequest.SMB2QueryDirectoryFlags.class);
+            }
 
-            @Override
-            public void remove() {
-                throw new UnsupportedOperationException("remove");
+            SMB2QueryDirectoryRequest qdr = new SMB2QueryDirectoryRequest(connection.getNegotiatedProtocol().getDialect(),
+                session.getSessionId(), treeConnect.getTreeId(),
+                getFileId(), FileInformationClass.FileIdBothDirectoryInformation,
+                flags,
+                0, null);
+
+            SMB2QueryDirectoryResponse qdResp;
+            try {
+                Future<SMB2QueryDirectoryResponse> qdFuture = session.send(qdr);
+                qdResp = Futures.get(qdFuture, TransportException.Wrapper);
+            } catch (TransportException e) {
+                throw new SMBRuntimeException(e);
             }
 
-            @Override
-            public FileInfo next() {
-                try {
-                    queryIfNeeded();
-                    FileInfo next = ctx.next;
-                    ctx.iteratorIndex++;
-                    return next;
-                } catch (TransportException e) {
-                    throw new SMBRuntimeException(e);
+            NtStatus status = qdResp.getHeader().getStatus();
+
+            if (status == NtStatus.STATUS_NO_MORE_FILES) {
+                return null;
+            } else {
+                if (status != NtStatus.STATUS_SUCCESS) {
+                    throw new SMBApiException(status, qdResp.getHeader().getStatusCode(), SMB2MessageCommandCode.SMB2_QUERY_DIRECTORY,
```

Comment:
Prefer to use the `SMBApiException(SMB2Header header, String message)` constructor. That one should work here :)